### ISSUE TITLE
Support the use of variables inside of if conditional blocks

### ISSUE THREAD

**Issue by jbroudou** [accordproject/cicero#549]
**Is your feature request related to a problem? Please describe.**
When using variables inside of an if conditional block the parser will return an error

```{{#if` isDelivered}} Delivered on {{deliveredAt}} {{/if}} ```

At the moment this is producing the error "Unexpected varid token: "deliveredAt""

**Describe the solution you'd like**
Allow the use of variables inside of if conditional blocks.

**Describe alternatives you've considered**
None exist as far as I know apart from not using an if conditional and changing deliveredAt to a String type with some sort of place holder text while isDelivered is false.

**Comment by dselman** [comment]
Thinking about this some more, perhaps as a workaround you could use an Ergo block?
https://docs.accordproject.org/docs/markup-ergo.html

That would allow you to conditionally insert text into the contract?

**Comment by jeromesimeon** [comment]
> **Is your feature request related to a problem? Please describe.**
> When using variables inside of an if conditional block the parser will return an error
> 
> `` {{#if` isDelivered}} Delivered on {{deliveredAt}} {{/if}} ``
> 
> At the moment this is producing the error "Unexpected varid token: "deliveredAt""
> 
> **Describe the solution you'd like**
> Allow the use of variables inside of if conditional blocks.
> 
> **Describe alternatives you've considered**
> None exist as far as I know apart from not using an if conditional and changing deliveredAt to a String type with some sort of place holder text while isDelivered is false.

This also possibly relates to #521.

@jbroudou Could you please share the model that you are expecting?

Currently `{{#if}}...{{/if}}` blocks correspond to Boolean variables. So let's say you have a model like this:
```
asset MyClause extends AccordClause {
  o Boolean isDelivered
  o DateTime deliveredAt
}
```
When `isDelivered` is `true` you would get the `deliveredAt` value, but what do you expect the contract data to be if `isDeliever` is `false` ?

**Comment by jeromesimeon** [comment]
Also this is somewhat of a duplicate of #17

**Comment by jbroudou** [comment]
@jeromesimeon The contract is a simple proof of concept
```
asset InspectionContract extends AccordContract {
  o AccordParty buyer
  o AccordParty seller
  o Boolean isDelivered
  o DateTime deliveredAt optional
  o DateTime inspectionCutOff optional
}```

When isDelivered is false I would expect deliveredAt to have no value and the template would not generate the statement that includes deliveredAt

**Comment by jeromesimeon** [comment]
A similar feature is supported in the upcoming new parser using the notion of optional block:
```
bash-3.2$ more grammar.tem.md 
This is a contract between {{seller}} and {{buyer}} for the amount of {{amount}} {{currency}}.{{#optional forceMajeure}}
This applies even in the presence of force majeure with a discount of {{rate}}%.{{/optional}}
bash-3.2$ more sample.md 
This is a contract between "Steve" and "Betty" for the amount of 3131.0 EUR.
This applies even in the presence of force majeure with a discount of 10.5%.
bash-3.2$ more model.cto 
namespace org.test

import org.accordproject.cicero.contract.* from https://models.accordproject.org/cicero/contract.cto
import org.accordproject.money.CurrencyCode from https://models.accordproject.org/money@0.2.0.cto

/**
 * The template model
 */
concept DiscountRate {
  o Double rate
}
asset MyClause extends AccordClause {
  o String seller
  o String buyer
  o Double amount
  o CurrencyCode currency
  o DiscountRate forceMajeure optional
}
bash-3.2$ ~/git/markdown-transform/packages/markdown-cli/index.js transform --to data --grammar grammar.tem.md --ctoFiles model.cto --input sample.md 
5:17:14 PM - info:
{
  "$class": "org.test.MyClause",
  "seller": "Steve",
  "buyer": "Betty",
  "amount": 3131,
  "currency": "EUR",
  "forceMajeure": {
    "$class": "org.test.DiscountRate",
    "rate": 10.5
  },
  "clauseId": "049f70d6-feff-4c11-99df-d43076d3dff3"
}
bash-3.2$ 
```

@jbroudou Would that address your use case?

**Comment by jbroudou** [comment]
@jeromesimeon I'm assuming that the optional block in your example will only render when forceMajeure has a "truthy" value? In which case, yes that would address my use case.

**Comment by jeromesimeon** [comment]
hi @jbroudou No this is based on the presence or absence of the optional value. There is no easy way to correlate the truthful value of one field to another.

Concretely, I do not know how to handle the cases:
- `isDelivered=true` but `deliveredAt=null` 
- `isDevlivered=false` but `deliveredAt="somedate"`

So I am not sure we would want to support that use case as is.

Suggestions for how to handle it or any feedback welcome.

**Comment by jbroudou** [comment]
@jeromesimeon I've had a look at [#17](https://github.com/accordproject/cicero/issues/17). Can you clarify what you mean by present or absent?

* DiscountRate forceMajeure optional

In this example how would forceMajeure be absent?

**Comment by jeromesimeon** [comment]
> @jeromesimeon I've had a look at [#17](https://github.com/accordproject/cicero/issues/17). Can you clarify what you mean by present or absent?
> 
> * DiscountRate forceMajeure optional
> 
> In this example how would forceMajeure be absent?

I will use the `forceMajeure` example above. For optional fields, it can either be present, with the following JSON serialization:
```
{
  "$class": "org.test.MyClause",
  "seller": "Steve",
  "buyer": "Betty",
  "amount": 3131,
  "currency": "EUR",
  "forceMajeure": {
    "$class": "org.test.DiscountRate",
    "rate": 10.5
  },
  "clauseId": "049f70d6-feff-4c11-99df-d43076d3dff3"
}
```

Or absent, with the following JSON serialization:
```
{
  "$class": "org.test.MyClause",
  "seller": "Steve",
  "buyer": "Betty",
  "amount": 3131,
  "currency": "EUR",
  "forceMajeure": null,
  "clauseId": "049f70d6-feff-4c11-99df-d43076d3dff3"
}```

**Comment by jbroudou** [comment]
Thanks for the clarification. This would indeed address my use case.

**Comment by jeromesimeon** [comment]
Great. Thanks for the feedback. This feature should go out with the next Cicero release.

**Comment by jeromesimeon** [comment]
Support for optionals is now available in Cicero `0.21`.
